### PR TITLE
Feature recommender beta

### PR DESCRIPTION
**Pull request by dattranm**

### CODE REVIEW

**Comment by ziedbouf** [comment]
any description to help understand the pull request, also is there any naming convention part of the project as i see different naming convention than the required by python

**Comment by dattranm** [comment]
Hi @ziedbouf , this part will be the Feature Recommender part, and it will soon be added as a part of ANOVOS package. It will contain several functions, and will use Semantic Text Similarity Model to recommend features to the users based on their input and interest. The details and demo of each function will be pushed later this week. Once again thank you so much for your interest.

**Review by natalie-lam — COMMENTED** [review]
- Standardise the function names, can change all to be separated by underscore (e.g., `list_all_industry, list_all_usecase, list_all_pair, list_usercase_by_industry, init_feature_recommender` etc.) 

- find instead of fine for `fine_attrByRelevance`? 

- Have another script for Sankey visualisation? 

- Move function explanations from notebook to the script as docstrings 

- For `feature_recommendation` function: 

   - Why not include the length of list_train in the `TypeError` print statement so user will know the limit for `top_n` 

   - Why not also include the `threshold` argument so users can tune both `top_n` and `threshold`

- For both `feature_recommendation` and `fine_attrByRelevance` functions, standardise code for `df_append`

   - In `feature_recommendation`, `df_append` is created by feeding in list of column values + column names 

   - In  `fine_attrByRelevance`, `df_append` is first initialised as empty dataframe with column names and the dictionary is used to fill in column values 

- For both `feature_recommendation` and `fine_attrByRelevance` functions, can specify the column names for the dataframes at the start so that the same list chunk does not need to be repeated twice 

- Print statement for `processIndustry` and `processUsecase` functions should be changed to: 

   - For `processIndustry` 

      - `print("Input Industry not available. Showing closest Industry semantic result for " , industry, ": ", processed_industry)` 

   - For `processUsecase` 

      - `print("Input Usecase not available. Showing closest Usecase semantic result for " , usecase, ": ", processed_usecase)`

**Comment by dattranm** [comment]
@natalie-lam Thank you for your comment and review.

> 1. Standardise the function names, can change all to be separated by underscore (e.g., `list_all_industry, list_all_usecase, list_all_pair, list_usercase_by_industry, init_feature_recommender` etc.)

Standardized the name of all the functions as suggested. In the latest push, I also split the Feature Exploration and Recommendation functions as well.
> 2. find instead of fine for `fine_attrByRelevance`?

Will keep the name, 'fine' is in fine tune Attribute by Relevance here
> 3. Have another script for Sankey visualisation?

Added Sankey visualization to be one of Feature Recommendation functions
> 4. Move function explanations from notebook to the script as docstrings

Will include
> 5. For `feature_recommendation` function:
>    
>    * Why not include the length of list_train in the `TypeError` print statement so user will know the limit for `top_n`
>    * Why not also include the `threshold` argument so users can tune both `top_n` and `threshold`

`TypeError` statement for `top_n` regarding to list_train already existed in the previous commit
Included `threshold` argument in `feature_recommendation` as suggested
> 6. For both `feature_recommendation` and `fine_attrByRelevance` functions, standardise code for `df_append`
>    
>    * In `feature_recommendation`, `df_append` is created by feeding in list of column values + column names
>    * In  `fine_attrByRelevance`, `df_append` is first initialised as empty dataframe with column names and the dictionary is used to fill in column values
> 7. For both `feature_recommendation` and `fine_attrByRelevance` functions, can specify the column names for the dataframes at the start so that the same list chunk does not need to be repeated twice

As this does not affect the code logic, will look into after resolving all the issues
> 8. Print statement for `processIndustry` and `processUsecase` functions should be changed to:
>    
>    * For `processIndustry`
>      
>      * `print("Input Industry not available. Showing closest Industry semantic result for " , industry, ": ", processed_industry)`
>    * For `processUsecase`
>      
>      * `print("Input Usecase not available. Showing closest Usecase semantic result for " , usecase, ": ", processed_usecase)`

Changed the output according to further discussion with @kajanansangar

**Review by natalie-lam — COMMENTED** [review]
I have reviewed the changes, all is good.

File: src/main/anovos/feature_recommender/feature_recommendation.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+                                 'Usecase', 'Source'])
+            df_out = df_out.append(df_append, ignore_index=True)
+    return df_out
+
+
+def find_attr_by_relevance(df, building_corpus, name_column=None, desc_column=None, threshold=0.3):
+    """
+    :param df: Input DataFrame - Attribute Dictionary
+    :param building_corpus: Input Feature Description (list)
+    :param name_column: Input column name for Attribute Name in Input DataFrame (string). Default is None.
```
Comment (kajanansangar) [inline src/main/anovos/feature_recommender/feature_recommendation.py:158]: Input, column name of Attribute Name in Input DataFrame (string). Default is None.

File: src/main/anovos/feature_recommender/feature_recommendation.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    :param desc_column: Input column name for Attribute Description in Input DataFrame (string). Default is None.
```
Comment (kajanansangar) [inline src/main/anovos/feature_recommender/feature_recommendation.py:159]: Input, column name of Attribute Description in Input DataFrame (string). Default is None.

**Review by kajanansangar — CHANGES_REQUESTED** [review]
I have commented on the respective files. Requested for changes. Further, I have not reviewed the notebook since that needs to be changed. I will review post the change

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    matches = finditer('.+?(?:(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|$)', input)
+    for m in matches:
+        processed_input += str(m.group(0)) + str(' ')
+    return processed_input
+
+
+def recommendation_data_prep(df, name_column, desc_column):
+    """
+    :param df: Input DataFrame
+    :param name_column: Input column name in Input DataFrame (string)
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/feature_recommender/featrec_init.py:21]: This function is for general cleaning purpose. Will change to "Indicating the column name of Input DataFrame attribute/ feature name"

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    :param desc_column: Input column description in Input DataFrame (string)
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/feature_recommender/featrec_init.py:22]: This function is for general cleaning purpose. Will change to "Indicating the column name of Input DataFrame attribute/ feature description"

File: src/main/anovos/feature_recommender/feature_exploration.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
+import pandas as pd
+import numpy as np
+from sentence_transformers import SentenceTransformer
+from sentence_transformers import util
+
+model = SentenceTransformer('all-mpnet-base-v2')
+input_path = 'https://raw.githubusercontent.com/anovos/anovos/feature_recommender_beta/data/feature_recommender/flatten_fr_db.csv'
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/feature_recommender/feature_exploration.py:7]: As ANOVOS comes within a package through pip install, there will not be /data path to be included within the installation. Therefore, the link to dataset in https is needed.

File: src/main/anovos/feature_recommender/feature_exploration.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    :param usecase: Input usecase (string)
+    :param semantic: Input semantic (boolean) - Whether the input needs to go through semantic matching or not. Default is True.
+    :return: Processed Usecase(string)
+    """
+    if type(semantic) != bool:
+        raise TypeError('Invalid input for semantic')
+    if type(usecase) != str:
+        raise TypeError('Invalid input for usecase')
+    usecase = usecase.lower().strip()
+    usecase = usecase.replace("[^A-Za-z0-9 ]+", " ")
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/feature_recommender/feature_exploration.py:48]: As discussed, leaving number should be good here

File: src/main/anovos/feature_recommender/feature_recommendation.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+import plotly.graph_objects as go
+import random
+import matplotlib.pyplot as plt
+
+
+def feature_recommendation(df, name_column=None, desc_column=None, suggested_industry='all', suggested_usecase='all',
+                           semantic=True, top_n=2, threshold=0.3):
+    """
+    :param df: Input DataFrame - Attribute Dictionary
+    :param name_column: Input column name for Attribute Name in Input DataFrame (string). Default is None.
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/feature_recommender/feature_recommendation.py:12]: Using 
`:param name_column: Input, column name of Attribute Name in Input DataFrame (string). Default is None.`
`:param desc_column: Input, column name of Attribute Description in Input DataFrame (string). Default is None.`
As indicated below

**Comment by dattranm** [comment]
> I have commented on the respective files. Requested for changes. Further, I have not reviewed the notebook since that needs to be changed. I will review post the change

I have resolved all comments above. Can you review again and lmk if anything? @kajanansangar

**Review by kajanansangar — APPROVED** [review]
Reviewed and made some changes

**Comment by ionicsolutions** [comment]
@dattranm @kajanansangar Can we merge this PR today?

**Pull request merged**